### PR TITLE
User directories + minor changes

### LINKED ISSUE #5: User directories class
[System.Environment.SpecialFolder](https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-7.0) is very Windows-oriented (not only by names, but also by values, for example, there's no XDG Documents folder, because both `MyDocuments` and `Personal` are home). Creating a class that will contain common XDG directories, including the ones parsed from `~/.config/user-dirs.dirs`, would be useful, and will allow to better follow XDG specs where needed (for example, we could move Parabolic's temp dirs to cache).

### PR DESCRIPTION
**Pull request by fsobolev**
Marking this as draft because I didn't test anything yet (will do tomorrow), although the lib builds and so I hope the code is done.

`UserDirectories` is a class allowing to get paths for various user folders. Unlike `Environment.SpecialFolder` it's more Linux-oriented, but should work on Windows too.

| Name | Linux | Windows |
| --- | --- | --- |
| Home | `$HOME`; `/home/{UserName}` | `SpecialFolder.Personal` |
| Config | `$XDG_CONFIG_HOME`; `{Home}/.config` | `SpecialFolder.ApplicationData` |
| ApplicationConfig | `{Config}/{AppInfo.Name}` | `{Config}\{AppInfo.Name}` |
| Cache | `$XDG_CACHE_HOME`; `{Home}/.cache` | `SpecialFolder.LocalApplicationData`<sup>[1]</sup> |
| ApplicationCache | `{Cache}/{AppInfo.Name}` | `{Cache}\{AppInfo.Name}` |
| LocalData<sup>[2]</sup> | `$XDG_DATA_HOME`; `{Home}/.local/share` | `SpecialFolder.ApplicationData` |
| ApplicationLocalData | `{LocalData}/{AppInfo.Name}` | `{LocalData}\{AppInfo.Name}` |
| Runtime | `$XDG_RUNTIME_DIR`; `/run/user/$UID` | Exception |
| Desktop | `$XDG_DESKTOP_DIR` (env var first, then the value from config file) | `SpecialFolder.DesktopDirectory`<sup>[3]</sup> |
| Documents | `$XDG_DOCUMENTS_DIR` (env var first, then the value from config file) | `SpecialFolder.MyDocuments` |
| Downloads | `$XDG_DOWNLOAD_DIR` (env var first, then the value from config file) | Get path using `SHGetKnownFolderPath` from `shell32.dll` |
| Music | `$XDG_MUSIC_DIR` (env var first, then the value from config file) | `SpecialFolder.MyMusic` |
| Pictures | `$XDG_PICTURES_DIR` (env var first, then the value from config file) | `SpecialFolder.MyPictures` |
| PublicShare | `$XDG_PUBLICSHARE_DIR` (env var first, then the value from config file) | Exception |
| Templates | `$XDG_TEMPLATES_DIR` (env var first, then the value from config file) | `SpecialFolder.Templates` |
| Videos | `$XDG_VIDEOS_DIR` (env var first, then the value from config file) | `SpecialFolder.MyVideos` |

<sup>[1]</sup> LocalApplicationData seems to be what is expected to use for cache and other temporary files on Windows
<sup>[2]</sup> I'm not sure about the name, but I think this is the least confusing
<sup>[3]</sup> I don't fully understand the difference between Desktop and DesktopDirectory in SpecialFolder, but from examples that I found DesktopDirectory seems to be a preferred option

I also modified ConfigurationLoader. When working on Cavalier I figured out that if the config contains invalid data (double number instead of integer for example), the app just crashes. The changes should prevent this: if JSON failed to deserialize data for whatever reason, an empty json is written to config file, but if the file existed, a backup is created first.

Closes #5

### CODE REVIEW

**Comment by fsobolev** [comment]
More changes:
1. I read how we use Aura in projects and found it a bit confusing that we use Aura constructor that returns an object but then we don't need this object because there's Aura.Active, so I made the constructor private and added static Aura.Init. It will also check that there's only one Aura.
2. Added AuraException to use instead of generic exception.
3. My first time writing tests 😄 It was easier than I thought, heh. Currectly there are tests for Aura initialization and for all UserDirectories (all passing ofc, but I will check everything with real code later today).

**Comment by fsobolev** [comment]
Tested with Tagger, all I had to do is a small change in MainWindowController:

```
        Aura.Init("org.nickvision.tagger", "Nickvision Tagger", _("Tagger"), _("Tag your music"));
        if (Directory.Exists($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}"))
        {
            // Move config files from older versions and delete old directory
            try
            {
                foreach (var file in Directory.GetFiles($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}"))
                {
                    File.Move(file, $"{UserDirectories.ApplicationConfig}{Path.DirectorySeparatorChar}{Path.GetFileName(file)}");
                }
            }
            catch (IOException) { }
            Directory.Delete($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}", true);
        }
```

Code for moving old config changed because Aura creates `ApplicationConfig` directory in getter (I did this on purpose: if we want to get some older it means we need it to exists), so we can't use `Directory.Move` like we did before.

File: Nickvision.Aura/UserDirectories.cs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    /// </summary>
+    /// <param name="key">XDG key</param>
+    /// <returns>Path string</returns>
+    private static string GetXDGUserDirectory(string key)
+    {
+        if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key)))
+        {
+            return Environment.GetEnvironmentVariable(key)!;
+        }
+        ParseXDGConfig();
```
Comment (nlogozzo) [inline Nickvision.Aura/UserDirectories.cs:290]: Move the code from the method itself here as it's only used here.

File: Nickvision.Aura.Tests/Aura.Test.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+using Xunit.Abstractions;
+
+namespace Nickvision.Aura.Tests;
+
+public class AuraTest
```
Comment (nlogozzo) [inline Nickvision.Aura.Tests/Aura.Test.cs:5]: Since the class name is `AuraTest` could name the file `AuraTest.cs` instead of `Aura.Test.cs`

**Review by nlogozzo — CHANGES_REQUESTED** [review]
Overall looks really good to me! The tests are super call too!!

Just one minor thing for the`UserDirectories` below. Also, could you maybe make a github action workflow for running the tests?

**Review by nlogozzo — APPROVED** [review]

**Pull request merged**